src/runtime/crdt/crdt-entity.ts

```diff
@@ -0,0 +1,202 @@
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {ChangeType, CRDTChange, CRDTModel, CRDTTypeRecord, VersionMap} from './crdt.js';
+import {Referenceable, CRDTCollectionTypeRecord, CRDTCollection, CollectionOpTypes} from './crdt-collection.js';
+import {CRDTSingletonTypeRecord, CRDTSingleton, SingletonOpTypes} from './crdt-singleton.js';
+import {Dictionary} from '../hot.js';
+
+type Identified = Dictionary<Referenceable>;
+
+type RawEntity<S extends Identified, C extends Identified> = 
+{
+  singletons: S,
+  collections: {[P in keyof C]: Set<C[P]>}
+};
+
+type SingletonEntityData<S extends Identified> = {[P in keyof S]: CRDTSingletonTypeRecord<S[P]>['data']};
+type CollectionEntityData<S extends Identified> = {[P in keyof S]: CRDTCollectionTypeRecord<S[P]>['data']};
+
+type SingletonEntityModel<S extends Identified> = {[P in keyof S]: CRDTSingleton<S[P]>};
+type CollectionEntityModel<S extends Identified> = {[P in keyof S]: CRDTCollection<S[P]>};
+
+export type EntityData<S extends Identified, C extends Identified> = 
+  {
+    singletons: SingletonEntityData<S>,
+    collections: CollectionEntityData<C>,
+    version: VersionMap
+  };
+
+export type EntityInternalModel<S extends Identified, C extends Identified> = 
+  {
+    singletons: SingletonEntityModel<S>,
+    collections: CollectionEntityModel<C>,
+    version: VersionMap
+  };
+
+export enum EntityOpTypes {Set, Clear, Add, Remove}
+
+type SetOp<S, F extends keyof S> = {type: EntityOpTypes.Set, field: F, value: S[F], actor: string, clock: VersionMap};
+type AddOp<C, F extends keyof C> = {type: EntityOpTypes.Add, field: F, value: C[F], actor: string, clock: VersionMap};
+type RemoveOp<C, F extends keyof C> = {type: EntityOpTypes.Remove, field: F, value: C[F], actor: string, clock: VersionMap};
+export type EntityOperation<S, C> = 
+  SetOp<S, keyof S> |
+  {type: EntityOpTypes.Clear, field: keyof S, actor: string, clock: VersionMap} |
+  AddOp<C, keyof C> |
+  RemoveOp<C, keyof C>;
+
+export interface CRDTEntityTypeRecord<S extends Identified, C extends Identified> extends CRDTTypeRecord {
+  data: EntityData<S, C>;
+  operation: EntityOperation<S, C>;
+  consumerType: RawEntity<S, C>;
+}
+
+type EntityModel<S extends Identified, C extends Identified> = CRDTModel<CRDTEntityTypeRecord<S, C>>;
+type EntityChange<S extends Identified, C extends Identified> = CRDTChange<CRDTEntityTypeRecord<S, C>>;
+
+export class CRDTEntity<S extends Identified, C extends Identified> implements EntityModel<S, C> {
+  private model: EntityInternalModel<S, C>;
+  
+  constructor(singletons: SingletonEntityModel<S>, collections: CollectionEntityModel<C>) {
+    this.model = {singletons, collections, version: {}};
+  }
+
+  merge(other: EntityData<S, C>): {modelChange: EntityChange<S, C>, otherChange: EntityChange<S, C>} {
+    const singletonChanges = {};
```

Comment:
Can you add a type here? Also to the lists below?

Reply:
nope :( Typescript's types just don't work that well in this scenario. Happy to go through in excruciating detail with you next week if you want.

Reply:
hmm, maybe I can, but it would be a coercion.

Reply:
OK, after playing around with this for some time, my considered opinion is that it isn't worth adding types.

(1) you need to annotate all Object.key() calls with `as (keyof S)[]`
(2) you need to invent a new type just for this internal parameter (it's too complex to put inline)
(3) you need to coerce the variable to that type as an empty dictionary isn't actually of that type; so in fact the type is incorrect for most of the function.
(4) typeScript can't figure out that allOps == true implies every member of singletonChanges / collectionChanges has an ops member.

Typescript types are great until they aren't.

Reply:
:-/

That's a shame.